signature/mechanism_sigstore.go

```diff
@@ -0,0 +1,325 @@
+}
+
+func (s *sigstoreSigningMechanism) GenerateCertificate() error {
+	fulcioServer, err := url.Parse(fulcioServer())
+	if err != nil {
+		return errors.Wrap(err, "parsing Fulcio URL")
+	}
+
+	fClient := fulcioClient.New(fulcioServer)
+	if fClient == nil {
+		return errors.New("error initializing fulcio client")
+	}
+
+	// TODO: use debug log
+	fmt.Println("Generating ephemeral keys")
+	priv, err := generatePrivateKey()
+	if err != nil {
+		return errors.Wrap(err, "generating private key")
+	}
+
+	signer, err := signature.LoadECDSASignerVerifier(priv, crypto.SHA256)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Retrieving signed certificate")
+	cert, chain, err := getCertForOauthID(priv, fClient.Operations, oidcIssuer, oidcClientID)
+	if err != nil {
+		return errors.Wrap(err, "retrieving cert")
+	}
+
+	s.ECDSASignerVerifier = signer
+	s.cert = cert
+	s.chain = chain
+	return nil
+}
+
+func generatePrivateKey() (*ecdsa.PrivateKey, error) {
+	return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+}
+
+func getCertForOauthID(priv *ecdsa.PrivateKey, ops operations.ClientService, oidcIssuer string, oidcClientID string) (certPem, chainPem []byte, err error) {
+	pubBytes, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tok, err := oauthflow.OIDConnect(oidcIssuer, oidcClientID, "", oauthflow.DefaultIDTokenGetter)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Sign the email address as part of the request
+	h := sha256.Sum256([]byte(tok.Subject))
+	proof, err := ecdsa.SignASN1(rand.Reader, priv, h[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bearerAuth := httptransport.BearerToken(tok.RawString)
+
+	content := strfmt.Base64(pubBytes)
+	signedChallenge := strfmt.Base64(proof)
+	params := operations.NewSigningCertParams()
+	params.SetCertificateRequest(
+		&fulcioModels.CertificateRequest{
+			PublicKey: &fulcioModels.CertificateRequestPublicKey{
+				Algorithm: fulcioModels.CertificateRequestPublicKeyAlgorithmEcdsa,
+				Content:   &content,
+			},
+			SignedEmailAddress: &signedChallenge,
+		},
+	)
+
+	resp, err := ops.SigningCert(params, bearerAuth)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// split the cert and the chain
+	certBlock, chainPem := pem.Decode([]byte(resp.Payload))
+	certPem = pem.EncodeToMemory(certBlock)
+	return certPem, chainPem, nil
+}
+
+// Sign creates a signature using the payload input.
+// Fails with a SigningNotSupportedError if the mechanism does not support signing.
+func (s *sigstoreSigningMechanism) Sign(payload []byte) ([]byte, error) {
+	return s.SignMessage(bytes.NewReader(payload), options.WithContext(s.ctx))
+}
+
+func (s *sigstoreSigningMechanism) Upload(signature, payload []byte) (*models.LogEntryAnon, error) {
+	rekorClient, err := rekorClient.GetRekorClient(rekorServer())
+	if err != nil {
+		return nil, err
+	}
+	re := rekorEntry(payload, signature, s.cert)
+	returnVal := models.Rekord{
+		APIVersion: swag.String(re.APIVersion()),
```

Comment:
`swag.String` seems to only return a pointer to the input string.  Could we do this manually and get rid of adding `swag` as a new dependency.

Note that I am trying to avoid new dependencies unless urgently necessary to keep our footprint small; it'll add some extra work downstream packaging as well (e.g., Debian has to devendor all dependencies and put them into extra source packages).

Reply:
Well, https://github.com/sigstore/rekor/blob/main/pkg/generated/models/rekord.go depends on `swag` anyway, so I don’t think this would buy us anything.